[Woo POS][Local Catalog] Periodic incremental sync of catalog

samiuelson opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
WOOMOB-1478

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR implements periodic incremental synchronization of the local catalog in the WooCommerce POS system. The sync runs automatically every hour while the POS home screen is active.

⚠️ Do not merge until the parent branch is `trunk`

### Test Steps
<!-- Describe how to test your changes. Include only what’s needed for the reviewer to validate the behavior.
For new features, outline the main user flow or goal rather than every tap or click — the reviewer should be able to complete the flow naturally.
If applicable, mention key devices, scenarios, or edge cases to verify. -->

* Verify periodic incremental sync is not being scheduled in case Local Catalog is disabled for the site
* Verify periodic incremental sync is working when WooPosActivity is in foreground
* Verify periodic incremental sync is canceled after leaving WooPosActivity

💡 For manual testing you can adjust value behind `WooPosPeriodicSyncFacade.SYNC_INTERVAL` to a lower value.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
N/A

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Merge branch `trunk` and resolve conflicts

samiuelson committed:
Bind hourly incremental sync to WooPosActivity lifecycle

samiuelson committed:
Add unit test for WooPosPeriodicSyncFacadeTest

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/WooPosHomeViewModel.kt:
```diff
@@ -66,6 +74,15 @@ class WooPosHomeViewModel @Inject constructor(
+                delay(SYNC_INTERVAL_MS)
+                incrementalSync.execute(WooPosIncrementalSyncReason.PERIODIC_HOURLY)
+            }
+        }
```
Redesigned in 5408d8d

samiuelson committed:
Execute periodic sync before delay

The periodic sync is now executed at the beginning of the cycle, before the delay. This ensures that a sync is performed immediately when the periodic sync is started, rather than waiting for the initial delay period to elapse.

samiuelson commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/root/WooPosPeriodicSyncFacade.kt:
```diff
@@ -0,0 +1,52 @@
+        periodicSyncJob = owner.lifecycleScope.launch {
+            while (isActive) {
+                delay(SYNC_INTERVAL)
+                incrementalSync.execute(WooPosIncrementalSyncReason.PERIODIC_HOURLY)
```
Fixed: d0a5960e6a35d14251070abc85e010cc73f99888

samiuelson committed:
Remove tests

samiuelson committed:
Merge branch 'trunk' into woomob-1478-woo-poslocal-catalog-consider-doing-periodic-incremental

samiuelson committed:
Do not schedule periodic incremental sync if local catalog not supported

samiuelson committed:
Adjust sync interval

samiuelson committed:
Merge branch 'woomob-1566-woo-poslocal-catalog-ptr-runs-incremental-sync-even-on' into woomob-1478-woo-poslocal-catalog-consider-doing-periodic-incremental

samiuelson committed:
Check if local catalog is supported before starting periodic sync

Adds a check to ensure the local catalog is supported for the selected site before initiating the periodic incremental sync. This prevents unnecessary sync attempts when the feature is not enabled.

samiuelson committed:
Merge remote-tracking branch 'origin/woomob-1478-woo-poslocal-catalog-consider-doing-periodic-incremental' into woomob-1478-woo-poslocal-catalog-consider-doing-periodic-incremental

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/root/WooPosPeriodicSyncFacade.kt

samiuelson committed:
Merge branch 'woomob-1566-woo-poslocal-catalog-ptr-runs-incremental-sync-even-on' into woomob-1478-woo-poslocal-catalog-consider-doing-periodic-incremental

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/root/WooPosPeriodicSyncFacade.kt:
```diff
@@ -0,0 +1,60 @@
+            if (!isLocalCatalogSupported(site.siteId)) return@launch
+
+            while (isActive) {
+                incrementalSync.execute(WooPosIncrementalSyncReason.PERIODIC_HOURLY)
```
💬 If I understand this correctly, the sync will run whenever:
- the User enters POS - seems like a duplication of POS_HOME sync, at least according to logs.
- the POS come to foreground - I guess this is fine.
- the display wakes up - I guess also fine?
- the POS activity comes to foreground - e.g. upon connecting to a reader (feels a bit off, but I guess also fine?)

I'm not against doing the sync this often, however, want to double check this is the intended behavior.

samiuelson replied:
Good catch 👍 I updated the condition to check the last sync timestamps. In case the sync was done within the last 1h (e.g., successful payment, POS home), the periodic sync will be skipped. Commit: 46f52a3

samiuelson replied:
Given that the incremental sync is now based on timestamp check, I adjusted the interval of checking the sync status: 89b1f25. This will ensure the sync is not missed in some cases, e.g., incremental sync was done 30 minutes ago on a successful payment; 30 minutes later, the while loop iterates and skips the sync, delaying the next sync try by 1 hour. Let me know what you think. Commit: 89b1f25

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/root/WooPosPeriodicSyncFacade.kt:
```diff
@@ -0,0 +1,60 @@
+
+/**
+ * Lifecycle-aware facade that performs periodic incremental syncs while [WooPosActivity] is visible.
+ * The sync runs once per hour while the POS activity is in the foreground (between onResume and onPause).
```
📓 If my below comment is correct, I believe this in-code comment is a bit misleading -> it's run whenever onResume is called, and it's also run once per hour given the POS has been in the foreground. Right?

samiuelson replied:
I updated the kdoc: 04e4b66. Given changes to the logic in 46f52a3, and 04e4b66, I believe it's fine now.

malinajirka reviewed (changes requested):
Thanks @samiuelson!


I've left a couple questions. I also noticed that both HOME and PERIODIC syncs are performed before the POS has a chance to do a full initial sync:

<img width="1988" height="288" alt="Screenshot 2025-10-29 at 12 17 54" src="https://github.com/user-attachments/assets/77313a42-d4c9-438f-a750-4c926eb06a28" />

samiuelson committed:
Conditionally trigger incremental sync based on last sync timestamp

Introduces a check to verify if an incremental sync is necessary before executing it. The sync will only be triggered if the time since the last product or variation sync is greater than or equal to the defined sync interval, or if no previous sync timestamp exists.

samiuelson committed:
Clarify WooPosPeriodicSyncFacade behavior comment

samiuelson committed:
Adjust interval of incremental sync status check

malinajirka reviewed (approved):
LGTM, thanks for the changes!
